### PR TITLE
Treat Some Warnings as Errors

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
See comments

### Testing Done
CI

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/cc_toolchain/cc_toolchain_config.bzl [hunk 1]
```diff
@@ -191,8 +191,7 @@ def _make_common_features(ctx):
                 actions = [ACTION_NAMES.c_compile, ACTION_NAMES.cpp_compile],
                 flag_groups = [
                     flag_group(
-                        flags = ["-Wall", "-Wextra", "-Wvla", "-Wconversion"] +
-                                ctx.attr.host_compiler_warnings,
+                        flags = ["-Wall", "-Werror"] + ctx.attr.host_compiler_warnings,
```
Comment (garethellis0) [inline src/cc_toolchain/cc_toolchain_config.bzl]: Why not have `-Wvla` and `-Wconversion`?

File: src/software/ai/passing/pass_generator_test.cpp [hunk 2]
```diff
@@ -122,7 +122,9 @@ TEST_F(PassGeneratorTest, check_pass_converges)
         EXPECT_LE((converged_pass.receiverPoint() - pass.receiverPoint()).length(), 0.3);
         EXPECT_LE(abs(converged_pass.speed() - pass.speed()), 0.3);
         EXPECT_LE(abs((converged_pass.startTime() - pass.startTime()).getSeconds()), 0.2);
+        (void)score;
     }
+    (void)converged_score;
```
Comment (garethellis0) [inline src/software/ai/passing/pass_generator_test.cpp]: Is this just ignoring unused variables? Could we perhaps add an `UNUSED` macro?

File: src/software/handheld_controller/controller.h [hunk 3]
```diff
@@ -121,7 +121,7 @@ class Controller : public Subject<ControllerInput>
     size_t getNumButtons(int fd) const;
 
     std::thread event_loop_thread;
-    std::atomic_bool in_destructor;
     std::shared_ptr<const HandheldControllerInputConfig> controller_input_config;
+    std::atomic_bool in_destructor;
```
Comment (garethellis0) [inline src/software/handheld_controller/controller.h:125]: nit; extra space below line 123

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Super happy to see this :smile:

File: src/software/ai/passing/pass_generator_test.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/passing/pass_generator_test.cpp]: done

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: * conversion causes errors in external libraries, so hard to fix
* vla causes errors in simulation/physics, where we have variable length arrays, which is hard to fix

File: src/software/handheld_controller/controller.h (hunk 3 above)
Comment (jonathanlew) [reply src/software/handheld_controller/controller.h:125]: I don't see it

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: actually, I can disable warnings at a more granular level

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: I have disabled errors on specific warnings that cause errors with libraries

Comment (MathewMacDougall) [reply src/cc_toolchain/cc_toolchain_config.bzl]: Just to be clear (for the physics stuff), you mean variable length arrays that are internal to Box2D or code we own? If it's the later I can try fix it.

**Review by MathewMacDougall — DISMISSED** [review]
:fire: :fire: :heart_eyes:

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: it's the latter, do you want to fix it as part of this PR?

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]:
```
In file included from firmware/shared/physics_linear_algebra.c:1:0:
./firmware/shared/physics_linear_algebra.h:13:18: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
                  const float matrix[lhs_len][rhs_len]);
                  ^~~~~
./firmware/shared/physics_linear_algebra.h:13:18: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
./firmware/shared/physics_linear_algebra.h:25:20: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
                    const float matrix[lhs_len][rhs_len]);
                    ^~~~~
./firmware/shared/physics_linear_algebra.h:25:20: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
firmware/shared/physics_linear_algebra.c:4:18: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
                  const float matrix[lhs_len][rhs_len])
                  ^~~~~
firmware/shared/physics_linear_algebra.c:4:18: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
firmware/shared/physics_linear_algebra.c:17:20: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
                    const float matrix[rhs_len][lhs_len])
                    ^~~~~
firmware/shared/physics_linear_algebra.c:17:20: error: ISO C90 forbids variable length array 'matrix' [-Werror=vla]
```

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]:
```
software/simulation/physics/physics_robot_model.cpp:53:40: error: variable length array 'vertices' is used [-Werror=vla]
     b2Vec2 vertices[shape_points.size()];
                                        ^
software/simulation/physics/physics_robot_model.cpp:60:43: warning: conversion to 'int32 {aka int}' from 'std::vector<Point>::size_type {aka long unsigned int}' may alter its value [-Wconversion]
     shape->Set(vertices, shape_points.size());
                          ~~~~~~~~~~~~~~~~~^~
software/simulation/physics/physics_robot_model.cpp: In static member function 'static b2PolygonShape* PhysicsRobotModel::getRobotBodyShapeFrontRight(double)':
software/simulation/physics/physics_robot_model.cpp:81:40: error: variable length array 'vertices' is used [-Werror=vla]
     b2Vec2 vertices[shape_points.size()];
                                        ^
software/simulation/physics/physics_robot_model.cpp:88:43: warning: conversion to 'int32 {aka int}' from 'std::vector<Point>::size_type {aka long unsigned int}' may alter its value [-Wconversion]
     shape->Set(vertices, shape_points.size());
                          ~~~~~~~~~~~~~~~~~^~

```

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]:
```
software/backend/output/radio/mrf/send_reliable_message_operation.cpp:24:34: error: variable length array 'buffer' is used [-Werror=vla]
         uint8_t buffer[3 + length];
                                  ^
```

Comment (MathewMacDougall) [reply src/cc_toolchain/cc_toolchain_config.bzl]: I'm fine fixing the simulation physics as part of this PR.

However @garethellis0 @EvanMorcom there was talk about possibly later using VLAs in the trajectory planner. Seems like these errors would prevent that. Is that something we'd really want to do later / see benefit in?

VLAs are actually something I thought we might use more for C stuff, but if it's going to fail when built with C++ we should think ahead about if we want to use VLAs at all or not.

**Review by akhilveeraghanta — DISMISSED** [review]
🚒

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (garethellis0) [reply src/cc_toolchain/cc_toolchain_config.bzl]: We can specify different flags for C and C++ IIRC.

File: src/software/ai/hl/stp/play/halt_play_test.cpp [hunk 4]
```diff
@@ -34,6 +34,7 @@ TEST(StopPlayTest, test_stop_play_returns_correct_tactics)
         {
             StopTactic *unused;
             unused = dynamic_cast<StopTactic *>(t.get());
+            UNUSED(unused);
```
Comment (garethellis0) [inline src/software/ai/hl/stp/play/halt_play_test.cpp:37]: So I'm not sure how this even compiles, but it seems like we're using the `UNUSED` macro from `test_util` here?

File: src/software/handheld_controller/controller.h (hunk 3 above)
Comment (garethellis0) [reply src/software/handheld_controller/controller.h:125]: Whoops, not sure what I was on, I don't see it either.........

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (garethellis0) [reply src/cc_toolchain/cc_toolchain_config.bzl]: With regards to the external libraries, I think having something is better then nothing, so if we need to ignore conversion warnings for now it's fine, let's just leave a comment why we are.

Comment (MathewMacDougall) [reply src/cc_toolchain/cc_toolchain_config.bzl]: Fair enough. @jonathanlew can you make a quick ticket to fix our internal uses of VLAS on the C++ side? (which I think is just the simulation physics and mrf stuff since we can handle the firmware stuff with C/C++ specific flags). Should be quick to fix afterwards and then we can re-enable the flag.

File: src/software/ai/hl/stp/play/halt_play_test.cpp (hunk 4 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/halt_play_test.cpp:37]: yes from test_util, do you have a better suggestion?

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: c/c++ specific flags in bazel? Do I put different flags in different toolchains?

Comment (garethellis0) [reply src/cc_toolchain/cc_toolchain_config.bzl]: Nah, you just need to have a `c_warnings_feature` and a `cpp_warnings_feature`, and then change the `actions` in each to apply to C/C++. Or even better have a `warnings_feature` that applies to both C/C++ and then a `cpp_warnings_feature` that just adds the vla warnings.

**Comment by jonathanlew** [comment]
Update:
* disabling warnings that cause errors in external libraries
* Enabled vla for cpp and made changes to robot model and radio code @MathewMacDougall, requesting a **MatCI**:tm: to make sure I didn't break radio
* disabled warnings for stm32 firmware because legacy_robot_stm32f4 has warnings

File: src/cc_toolchain/cc_toolchain_config.bzl [hunk 5]
```diff
@@ -722,7 +741,7 @@ def _stm32_impl(ctx):
             "c++17",
             "colour",
             "determinism",
-            "warnings",
+            #TODO: enable warning_as_errors when we remove legacy_robot_stm32f4
```
Comment (MathewMacDougall) [inline src/cc_toolchain/cc_toolchain_config.bzl]: Can this be linked to a ticket?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

**Comment by MathewMacDougall** [comment]
MatCI :tm: passed. Radio works :+1:

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 5 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: yup

File: src/software/simulation/physics/box2d_util.cpp [hunk 6]
```diff
@@ -54,7 +54,7 @@ float polygonArea(const b2PolygonShape& polygon)
     // counter-clockwise.
     double area    = 0.0;
     unsigned int j = polygon.m_count - 1;
-    for (unsigned int i = 0; i < polygon.m_count; i++)
+    for (int i = 0; i < polygon.m_count; i++)
```
Comment (EvanMorcom) [inline src/software/simulation/physics/box2d_util.cpp:57]: Why was this change needed?

**Review by EvanMorcom — DISMISSED** [review]
LGTM 🐕‍🦺 🦮 

Just one question.

File: src/software/simulation/physics/box2d_util.cpp (hunk 6 above)
Comment (jonathanlew) [reply src/software/simulation/physics/box2d_util.cpp:57]: m_count is an int, I don't know why :thinking: @MathewMacDougall

File: src/software/ai/hl/stp/play/halt_play_test.cpp (hunk 4 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/play/halt_play_test.cpp:37]: Oh sorry, thought this was `halt_play` not `halt_play_test` (hence how it's able to compile), my bad!

**Review by garethellis0 — DISMISSED** [review]
Doing gods work @jonathanlew. 
LGTM :+1: :tada:

**Review by kelvinkoon — DISMISSED** [review]

File: src/software/simulation/physics/box2d_util.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/software/simulation/physics/box2d_util.cpp:57]: Box2D just do be like that :shrug:

**Review by MathewMacDougall — DISMISSED** [review]

**Review by akhilveeraghanta — DISMISSED** [review]

**Comment by jonathanlew** [comment]
Current state of affairs: 
* toolchain defaults to -Werror
* Add -Wall and -Wextra to tbots packages with unused_parameter disabled
* Override -Werror with copts=-Wnoerror on specific targets that we don't want to have errors, e.g. 3rd party code and legacy code

We can add -Wconversion later (in another issue)

Comments on this structure?

File: src/cc_toolchain/cc_toolchain_config.bzl [hunk 7]
```diff
@@ -184,14 +184,14 @@ def _make_common_features(ctx):
         enabled = True,
     )
 
-    result["warnings_feature"] = feature(
-        name = "warnings",
+    result["warnings_as_errors_feature"] = feature(
```
Comment (MathewMacDougall) [inline src/cc_toolchain/cc_toolchain_config.bzl:187]: Just to make sure I'm understanding this, we still need the `warnings_as_errors_feature` to actually turn warnings into errors, and everything in the `bazelrc` is just determining what warnings we have enabled?

**Review by MathewMacDougall — COMMENTED** [review]
I like the new structure :+1:

**Review by kelvinkoon — DISMISSED** [review]
New structure LGTM! :+1:

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 7 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl:187]: Yes, I can't have -Werror in the bazelrc, because it can't be overridden by the `copts=["-Wnoerror"]` in the legacy and 3rd party code, it has to live in the toolchain unfortunately.

Comment (MathewMacDougall) [reply src/cc_toolchain/cc_toolchain_config.bzl:187]: I think that's fine, just wanted to make sure I understood.

Comment (MathewMacDougall) [reply src/cc_toolchain/cc_toolchain_config.bzl:187]: Can you maybe add that as a comment somewhere just to be super clear to future build-systems people?

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl:187]: added comment

**Review by akhilveeraghanta — DISMISSED** [review]
very nice

**Review by MathewMacDougall — DISMISSED** [review]
Gods work :pray: :fire: so glad this is being done. Nice work! :tada:

File: src/.bazelrc [hunk 8]
```diff
@@ -9,6 +9,12 @@ test --test_output=all
 # Use Our Custom Toolchain
 build --crosstool_top=//cc_toolchain:toolchain
 
+# Add warnings to Thunderbots code only
+build --per_file_copt=//software/.*,//shared/.*,//firmware/.*,//firmware_new/.*@-Wall,-Wextra,-Wno-unused-parameter
```
Comment (garethellis0) [inline src/.bazelrc:13]: Can we blacklist the anything that matches `@//` and the `external` folder instead of whitelisting our folders?

**Review by garethellis0 — COMMENTED** [review]
nice, this is slick now

File: src/.bazelrc (hunk 8 above)
Comment (jonathanlew) [reply src/.bazelrc:13]: I tried that, --per_file_copt doesn't work on the @zlib//:zlib dependency of protobuf and I don't know why. Even .* (as blacklist or whitelist) won't capture zlib, and of course zlib throws warnings.

**Review by EvanMorcom — DISMISSED** [review]
LGTM 🔥

**Comment by jonathanlew** [comment]
Fixed merge conflicts, please re-review

**Review by MathewMacDougall — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]

File: src/.bazelrc (hunk 8 above)
Comment (garethellis0) [reply src/.bazelrc:13]: OK, thanks for the context.

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:
